iio/iio.c

```diff
@@ -0,0 +1,838 @@
+ */
+static int16_t iio_get_channel_id(const char *channel,
+				  struct iio_channel **channels)
+{
+	int16_t i = 0;
+
+	if (!(*channels))
+		return -EINVAL;
+
+	while (channels[i]) {
+		if (!strcmp(channel, channels[i]->name))
+			return i;
+		i++;
+	}
+
+	return -ENOENT;
+}
+
+/**
+ * @brief Get attribute ID from a list of attributes.
+ * @param *attr - Attribute name.
+ * @param *attributes - List of attributes.
+ * @return attribute ID, or negative value if attribute is not found.
+ */
+static int16_t iio_get_attribute_id(const char *attr,
+				    struct iio_attribute **attributes)
+{
+	int16_t i = 0;
+
+	if (!(*attributes))
+		return -EINVAL;
+
+	while (attributes[i]) {
+		if (!strcmp(attr, attributes[i]->name))
+			return i;
+		i++;
+	}
+
+	return -ENOENT;
+}
+
+/**
+ * @brief Find interface with "device_name".
+ * @param @device_name - Device name.
+ * @param iio_interfaces - List of interfaces.
+ * @return interface pointer if interface is found, NULL otherwise.
+ */
+static struct iio_interface *iio_get_interface(const char *device_name,
+		struct iio_interfaces *iio_interfaces)
+{
+	int16_t i;
+
+	for (i = 0; i < iio_interfaces->num_interfaces; i++) {
+		if (!strcmp(device_name, iio_interfaces->interfaces[i]->name))
+			return iio_interfaces->interfaces[i];
+	}
+
+	return NULL;
+}
+
+/**
+ * @brief Read all attributes from an attribute list.
+ * @param *device - Physical instance of a device.
+ * @param *buf - Buffer where values are read.
+ * @param len - Maximum length of value to be stored in buf.
+ * @param *channel - Channel properties.
+ * @param **attributes - List of attributes to be read.
+ * @return number of bytes read.
+ */
+static ssize_t iio_read_all_attr(void *device, char *buf, size_t len,
+				 const struct iio_ch_info *channel, struct iio_attribute **attributes)
+{
+	int16_t i = 0, j = 0;
+	char local_buf[256];
+	ssize_t attr_length;
+	uint32_t *pattr_length;
+
+	while (attributes[i]) {
+		attr_length = attributes[i]->show(device, local_buf, len, channel);
+		pattr_length = (uint32_t *)(buf + j);
+		*pattr_length = bswap_constant_32(attr_length);
+		j += 4;
+		if (attr_length >= 0) {
+			sprintf(buf + j, "%s", local_buf);
+			if (attr_length & 0x3) /* multiple of 4 */
+				attr_length = ((attr_length >> 2) + 1) << 2;
+			j += attr_length;
+		}
+		i++;
+	}
+
+	return j;
+}
+
+/**
+ * @brief Write all attributes from an attribute list.
+ * @param *device - Physical instance of a device.
+ * @param *buf - Values to be written.
+ * @param len - Length of buf.
+ * @param *channel - Channel properties.
+ * @param **attributes - List of attributes to be written.
+ * @return number of written bytes.
+ */
+static ssize_t iio_write_all_attr(void *device, char *buf, size_t len,
+				  const struct iio_ch_info *channel, struct iio_attribute **attributes)
+{
+	int16_t i = 0, j = 0;
+	int16_t attr_length;
+
+	while (attributes[i]) {
+		attr_length = bswap_constant_32((uint32_t)(buf + j));
```

Comment:
`attributes` could be NULL, and I think a NULL `buf` could also create issues

Reply:
Fixed